### PR TITLE
fix: strip markdown code fences before JSON parsing in structured output

### LINKED ISSUE #959: [BUG] v1.12.0: JSON parse error with Claude models via OpenAI-compatible endpoints (LiteLLM/OpenRouter)
## Description

After upgrading to v1.12.0, Claude models (e.g. `claude-haiku-4.5`) fail with JSON parse errors when used via OpenAI-compatible endpoints like LiteLLM or OpenRouter. The same setup worked perfectly in v1.11.x.

## Error
```
SyntaxError: Unexpected token '`', "```json
{
"... is not valid JSON
    at JSON.parse (<anonymous>)
    at async p.generateObject (...)
```

## Root Cause

The migration from LangChain to Vercel AI SDK in v1.12.0 introduced this regression. The `generateObject()` function expects raw JSON, but Claude (and potentially other models) often wrap their JSON output in markdown code blocks when accessed via OpenAI-compatible proxies.

LangChain was tolerant of this. Vercel AI SDK's `generateObject()` is not.

## Workaround

Switching to Gemini models (e.g. `gemini-2.5-flash`) works - they return raw JSON without markdown wrapping.

## Suggested Fix

Strip markdown code block wrappers from LLM responses before parsing JSON.

## Environment

- Perplexica: v1.12.0 (slim image)
- Provider: LiteLLM (OpenAI-compatible endpoint)
- Chat model: claude-haiku-4.5

### PR DESCRIPTION
**Pull request by VibhorGautam**
## Summary

Fixes #959

Some LLM providers (notably Claude via OpenAI-compatible APIs and certain Ollama models) wrap their structured output in markdown code fences:

````
```json
{"key": "value"}
```
````

This causes `JSON.parse` and `partial-json` to throw parse errors, breaking `generateObject()` and `streamObject()` calls with these models.

## Changes

- Added a `stripMarkdownFences()` helper that removes `` ```json ... ``` `` wrappers before the content reaches `repairJson` / `parse`
- Applied to both **OpenAI** and **Ollama** provider implementations
- Covers both `generateObject()` (full response) and `streamObject()` (streaming partial + final)
- Other providers (Groq, Gemini, Anthropic, LMStudio, Lemonade) extend `OpenAILLM` so they inherit the fix automatically

## Why `repairJson({ extractJson: true })` isn't enough

The existing `repairJson` with `extractJson` handles some cases but doesn't reliably strip markdown fences in all edge cases — particularly when the fence includes a language tag (`json`) or has varying whitespace. The explicit strip before repair makes the pipeline more robust.

## How to test

1. `npm install && npm run dev`, open http://localhost:3000
2. In settings, pick a model that's known to wrap structured output in code fences (e.g. Claude via an OpenAI-compatible proxy, or an Ollama model like `deepseek-r1`)
3. Send a query that triggers the search pipeline (e.g. "latest news about TypeScript") — this exercises `generateObject()` in the classifier
4. Confirm the search completes without a JSON parse error in the terminal logs
5. Also test with a standard OpenAI model (e.g. GPT-4o) to make sure the no-fence path still works — `stripMarkdownFences` should be a no-op when there are no fences

### CODE REVIEW

File: src/lib/models/providers/ollama/ollamaLLM.ts [hunk 1]
```diff
@@ -13,6 +13,17 @@ import crypto from 'crypto';
...
 import { repairJson } from '@toolsycc/json-repair';
 
+/**
+ * Some models wrap their JSON output in markdown code fences like
+ * ```json\n{...}\n``` which breaks downstream parsing. This strips
+ * those fences so we get the raw JSON string.
+ */
+function stripMarkdownFences(text: string): string {
+  const trimmed = text.trim();
+  const match = trimmed.match(/^```(?:json)?\s*\n?([\s\S]*?)\n?\s*```$/);
```
[the remark this answers is not in the document]
Comment (VibhorGautam) [reply src/lib/models/providers/ollama/ollamaLLM.ts]: Good catch — fixed in b1b3c8b. The regex now has two stages: first tries to match a full fence pair, then falls back to stripping just a leading fence (no closing ``````` required). This handles the streaming case where the accumulated text starts with ``````json\n{``` but the closing fence hasn't arrived yet.

File: src/lib/models/providers/openai/openaiLLM.ts [hunk 2]
```diff
@@ -256,14 +269,14 @@ class OpenAILLM extends BaseLLM<OpenAIConfig> {
         recievedObj += chunk.delta;
 
         try {
-          yield parse(recievedObj) as T;
+          yield parse(stripMarkdownFences(recievedObj)) as T;
```
[the remark this answers is not in the document]
Comment (VibhorGautam) [reply src/lib/models/providers/openai/openaiLLM.ts:278]: Addressed in the same commit (b1b3c8b). The strip function now handles opening-only fences for streaming partials — the leading ``````json\n``` gets stripped even when there's no closing ```````` yet, so `parse()` receives clean partial JSON.

File: src/lib/models/providers/openai/openaiLLM.ts [hunk 3]
```diff
@@ -20,6 +20,23 @@ import {
...
+/**
+ * Some models wrap their JSON output in markdown code fences like
+ * ```json\n{...}\n```. This strips those fences so we get raw JSON.
+ * Also handles the streaming case where only the opening fence is
+ * present (the closing fence hasn't arrived yet).
+ */
+function stripMarkdownFences(text: string): string {
+  const trimmed = text.trim();
+  // Full fence pair: ```json\n...\n```
+  const full = trimmed.match(/^```(?:json)?\s*\n([\s\S]*?)\n?\s*```\s*$/);
```
[the remark this answers is not in the document]
Comment (VibhorGautam) [reply src/lib/models/providers/openai/openaiLLM.ts]: Fixed in c48cdd0 — changed `\n` to `\n?` in both the full-pair and leading-only regex patterns so same-line fenced JSON like ``````json{"key":"value"}``````` is handled correctly.
